spec/classes/stub_spec.rb

```diff
@@ -0,0 +1,275 @@
+
+    $stderr = stub_stderr_file
+    $stdout = stub_stdout_file
+    allow($stdin).to receive(:tty?).and_return(false)
+
+    allow_any_instance_of(Kernel).to receive(:exit) do |_, exit_code|
+      exit_code_list << exit_code
+      raise SystemExit
+    end
+  end
+
+  context 'with no configuration' do
+    let(:stub_call_log) do
+      ARGV.replace []
+      allow($stdin).to receive(:read).and_return('')
+      load subject_path
+
+      ARGV.replace []
+      allow($stdin).to receive(:read).and_return("dog\ncat\n")
+      load subject_path
+
+      ARGV.replace %w(first_argument second_argument)
+      allow($stdin).to receive(:read).and_return('')
+      load subject_path
+
+      ARGV.replace %w(first_argument second_argument)
+      allow($stdin).to receive(:read).and_return("dog\ncat\n")
+      load subject_path
+
+      YAML.load(stub_call_file.string)
+    end
+
+    it 'logs a blank STDIN for the first call' do
+      expect(stub_call_log[0]['stdin']).to be_empty
+    end
+
+    it 'logs no arguments for the first call' do
+      expect(stub_call_log[0]['args']).to be_nil
+    end
+
+    it 'logs some STDIN for the second call' do
+      expect(stub_call_log[1]['stdin']).to eql "dog\ncat\n"
+    end
+
+    it 'logs no arguments for the second call' do
+      expect(stub_call_log[1]['args']).to be_nil
+    end
+
+    it 'logs a blank STDIN for the third call' do
+      expect(stub_call_log[2]['stdin']).to be_empty
+    end
+
+    it 'logs some arguments for the third call' do
+      expect(stub_call_log[2]['args']).to eql %w(first_argument second_argument)
+    end
+
+    it 'logs some STDIN for the fourth call' do
+      expect(stub_call_log[3]['stdin']).to eql "dog\ncat\n"
+    end
+
+    it 'logs some arguments for the fourth call' do
+      expect(stub_call_log[3]['args']).to eql %w(first_argument second_argument)
+    end
+
+    it 'exits with appropriate code for first call' do
+      expect(exit_code_list[0]).to be_nil
+    end
+
+    it 'exits with appropriate code for second call' do
+      expect(exit_code_list[1]).to be_nil
+    end
+
+    it 'exits with appropriate code for third call' do
+      expect(exit_code_list[2]).to be_nil
+    end
+
+    it 'exits with appropriate code for fourth call' do
+      expect(exit_code_list[3]).to be_nil
+    end
+  end
+  context 'with configurations that does not end up matching calls' do
+    before(:each) do
+      stdout_configuration = [
+        {
+          args: ['super_special_argument'],
+          outputs: [
+            {
+              target: :stderr,
+              content: "no args content\n"
+            }
+          ],
+          statuscode: 1
+        }
+      ]
+      allow($stdin).to receive(:read).and_return('')
+      allow(stub_config_pathname).to receive(:exist?).and_return(true)
+      allow(stub_config_pathname).to receive(:read).and_return(stdout_configuration.to_yaml)
+
+      ARGV.replace []
+      begin
+        load subject_path
+      rescue SystemExit
```

Comment:
These try catches suck, but I struggled to come up with a better way of not letting the stub exits bomb out the entire test. Maybe the at_exit hook stuff would be a better alternative?